Added aws s3 data node

Forchapeatl opened the pull request:
### Added AWS S3 DataNode
![image](https://github.com/Avaiga/taipy-core/assets/24577149/46dfc721-ba8e-4d38-a038-7dc746e69ca7)

Fixes Avaiga/taipy#536
- [x] Implemented a new predefined storage type "s3" in DataNodeConfig, to model data stored on an S3 bucket.
- [x] List and design the configuration attributes required.
        -    Create S3 datanode.
        -    Read stream object from s3 bucket.
        -    Write  stream object  to s3 bucket.
- [x] Implemented the corresponding S3DataNode inheriting from the parent DataNode class.

### Todo
- [ ] Add reference manual documentation.
- [ ] Write file to s3 bucket

Forchapeatl committed:
Create aws_s3.py

Forchapeatl committed:
Create test_aws_s3_data_node.py

Forchapeatl committed:
Update Pipfile

Forchapeatl commented:
Hello @toan-quach  and @jrobinAV , please permit me to tag you directly. For some reason,  I can't seem to request for a review.
Please have a look at my PR.

toan-quach commented:
@Forchapeatl awesome!! Thanks for tagging us! We will review as soon as we can :D Thanks again for your help!

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+
+
+class AWSs3DataNode(DataNode):
+    """Data Node stored in a Amazon Web Service S3 Bucket .
```
```suggestion
    """Data Node stored in an Amazon Web Service S3 Bucket.
```

Forchapeatl replied:
>The configure method in the DataNodeConfig is missing though.

Thank you @jrobinAV ,please for more infomation on the` DataNodeConfig `method. I having diffuculties in understanding what is missing. It seems the method is required in the **test_aws_s3_data_node.py** file but I am not really sure of myself.

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+    __AWS_SECRET_ACCESS_KEY = "aws_secret_acces_key"
+    __AWS_REGION = "aws_region"
+    __AWS_STORAGE_BUCKET_NAME = "aws_s3_bucket_name"
+    __AWS_S3_OBJECT_PARAMETERS  = "aws _S3_object_parameters"
```
```suggestion
    __AWS_S3_OBJECT_PARAMETERS  = "aws_S3_object_parameters"
```

trgiangdo replied:
```suggestion
    __AWS_S3_OBJECT_PARAMETERS  = "aws_s3_object_parameters"
```

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+                f"The following properties " f"{', '.join(x for x in missing)} were not informed and are required"
+            )
+
+
```
```suggestion
```

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+        required = self._REQUIRED_PROPERTIES
+        if missing := set(required) - set(properties.keys()):
+            raise MissingRequiredProperty(
+                f"The following properties " f"{', '.join(x for x in missing)} were not informed and are required"
```
```suggestion
                f"The following properties " f"{', '.join(x for x in missing)} were not informed and are required."
```

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+
+
+
+
```
```suggestion
```

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+            **properties,
+        )
+
+        self.s3_client = boto3.client('s3',
```
In Taipy, we prefix names (classes, variables, methods, ...) by an `_` when we don't want to expose the variable to our users.

I would not expose the client, so I would call the variable `_s3_client`.

```suggestion
        self.s3_client = boto3.client('s3',
```

Forchapeatl replied:
> I would not expose the client, so I would call the variable `_s3_client`.

```
       self._s3_client = boto3.client('s3',
```

trgiangdo commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+
+
+
+class AWSs3DataNode(DataNode):
```
```suggestion
class S3DataNode(DataNode):
```
I would rename this for simplicity

trgiangdo commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+        editor_id (Optional[str]): The identifier of the user who is currently editing the data node.
+        editor_expiration_date (Optional[datetime]): The expiration date of the editor lock.
+        properties (dict[str, Any]): A dictionary of additional properties. Note that the
+            _properties_ parameter must at least contain an entry for _"db_name"_ and _"collection_name"_:
```
```suggestion
            _properties_ parameter must at least contain an entry for _"aws_access_key"_ and _"aws_secret_acces_key"_:
```

trgiangdo commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+            - _"aws_secret_acces_key"_ `(str)`: Amazon Web Services access key to authenticate programmatic requests.\n
+            - _"aws_region"_ `(Any)`: Self-contained geographic area where Amazon Web Services (AWS) infrastructure is located.\n
+            - _"aws_s3_bucket_name"_ `(str)`: unique identifier for a container that stores objects in Amazon Simple Storage Service (S3).\n
+            - _"aws _S3_object_parameters"_ `(str)`: A dictionary of additional arguments to be passed to interact with the AWS service\n
```
```suggestion
            - _"aws_s3_object_parameters"_ `(str)`: A dictionary of additional arguments to be passed to interact with the AWS service\n
```

trgiangdo commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+
+    """
+
+    __STORAGE_TYPE = "aws_s3_bucket"
```
```suggestion
    __STORAGE_TYPE = "aws_s3"
```

Small rename for simplicity

trgiangdo commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+        return cls.__STORAGE_TYPE
+
+
+    def read_object_stream(self ,bucket_name , object_key ):
+        aws_s3_object = self.s3_client.get_object(
+            Bucket = bucket_name,
+            Key = object_key
+        )
+        return  aws_s3_object['Body'].read().decode('utf-8')
```
In the `DataNode` parent class, we have 3 abstract method that need to be implemented for all children classes, including `_read()`, `_write()`, and `_append()` methods.

The `_read()` method can be implemented similar to this `read_object_stream()` here, however the bucket_name and object_key can be retrieve from `self.properties`.

This would mean the `__AWS_STORAGE_BUCKET_NAME` and `__AWS_S3_OBJECT_PARAMETERS` would be `_REQUIRED_PROPERTIES`.

Forchapeatl replied:
Thank you @trgiangdo , I can see the logic clearer now.

trgiangdo commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+        )
+        return  aws_s3_object['Body'].read().decode('utf-8')
+
+    def write_object_stream(self , bucket_name , object_key ,  data  ):
+        self.s3_client.put_object(
+            Bucket = bucket_name,
+            Key = object_key,
+            Body = data
+            )
```
Similarly, the `_write()` method can be implemented similar to this `write_object_stream()` here, however the bucket_name and object_key can be retrieve from self.properties.

However, this method seems to "append" data to the bucket. The `_write()` method should delete all data and rewrite.

You can also implemented the `_append()` method, however it's optional.

Forchapeatl replied:
I don't mind  including the append method, but  I can't really understand it's use case for S3 . Do you mean **adding**  more data to the **same** s3 bucket ?

trgiangdo replied:
Yes

trgiangdo commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+    __STORAGE_TYPE = "aws_s3_bucket"
+    __AWS_ACCESS_KEY_ID = "aws_access_key"
+    __AWS_SECRET_ACCESS_KEY = "aws_secret_acces_key"
+    __AWS_REGION = "aws_region"
```
Where is this `aws_region` used? Should be use when connecting to AWS right?

trgiangdo reviewed (commented):
Thank you for a great first PR. I believe this will be a good addition to taipy-core.

I have a few comments about naming and DataNode abstract methods that need to be implemented.
Other than that, there are minor linting errors that should be fixed.

Forchapeatl committed:
renamed variables , implemented abstract methods

Forchapeatl committed:
test S3DataNode and class methods

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+
+
+
+class S3DataNode(DataNode):
```
Since the data node only reads one object and not a whole bucket, I would change the name to something like:
S3ObjectDataNode or AwsS3ObjectDataNode.

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+        return cls.__STORAGE_TYPE
+
+
+    def _read(self , object_key ):
```
The Taipy Orchestrator uses the `_read` method. It is a generic way of reading the data. The orchestrator has no way to pass any parameter.

The method signature must be:
```
def _read(self):

```

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,156 @@
+        )
+        return  aws_s3_object['Body'].read().decode('utf-8')
+
+    def _write(self ,object_key ,  data  ):
```
Same comment. The `_write` method is used by the Taipy orchestrator when executing a Task. 
The signature must be: 
```
    def _write(self, data: Any):
```
That means the object_key must be passed to the data node at the creation, through the datanode configuration.

jrobinAV reviewed (commented):
The PR is going in the right direction. Thanks a lot.

The configure method in the DataNodeConfig is missing though.

jrobinAV commented:
Hello @Forchapeatl.

Thank you very much for your contribution. For information, we are currently working on a significant repository refactoring. We are merging sub-repositories into a single repository. For instance, the taipy-core repository is about to be merged into the taipy repository.

This should not impact your work.

But just so you are not surprised, instead of merging your PR, we will move your code from this branch (in taipy-core) into a new branch of another repository (taipy). We will let you know and let you review the code.

Thank you for your understanding.

Forchapeatl committed:
Merge branch 'Avaiga:develop' into Added-AWS-S3-DataNode

Forchapeatl committed:
Update aws_s3.py

Forchapeatl committed:
Switched  S3DataNode to S3ObjectDataNode

Forchapeatl committed:
Moved testing from S3DataNode to  S3objectDataNode

Forchapeatl commented:
Hello @jrobinAV thank you for your kind review. The requested changes have been made. The `S3ObjectDataNode` has been implemented  and tested with the `_read` and `_write` fuctions in thier expected signatures. 

The issue now is the `_append` method. I had a [chat](https://github.com/Avaiga/taipy-core/pull/828#discussion_r1398062266) with @trgiangdo concerning this  method.  I saw how this was possible when we were **adding files** to a **single** bucket.  Now our DataNode is making reference to a **single object** not **bucket**. Handling this method should be possible  by copying all the data to a new object, appendng the new content and writing back to S3 , but I don't think it's worth the effort in cases where the object data (file) is very large ( _I may be wrong_ ) . Please for your opinion.

[Reference: Append Data to S3 Object](https://stackoverflow.com/questions/41783903/append-data-to-an-s3-object)

trgiangdo commented:
Thank you for your information. `_append()` is an optional method, so if S3 does not support appending, then it's totally fine.

trgiangdo commented:
About the `DataNodeConfig`, you probably want to take a look at this [documentation page](https://docs.taipy.io/en/latest/manuals/core/concepts/).

Then, you can take a look at the `src/taipy/core/config/data_node_config.py` file, where we define different configuring methods for different data node types, and implement a new `configure_s3_data_node()` method.

For now, the todo list:
- [ ] Implement `configure_s3_data_node()` method to the `DataNodeConfig` class
- [ ] Add `_STORAGE_TYPE_VALUE_S3` attribute and add that attribute to the `_ALL_STORAGE_TYPES` list.
- [ ] Update `DataNodeConfig` docstring.
- [ ] Add configuration tests for S3DataNode that created from an s3 DataNodeConfig.

jrobinAV commented:
> Hello @jrobinAV thank you for your kind review. The requested changes have been made. The `S3ObjectDataNode` has been implemented and tested with the `_read` and `_write` fuctions in thier expected signatures.
> 
> The issue now is the `_append` method. I had a [chat](https://github.com/Avaiga/taipy-core/pull/828#discussion_r1398062266) with @trgiangdo concerning this method. I saw how this was possible when we were **adding files** to a **single** bucket. Now our DataNode is making reference to a **single object** not **bucket**. Handling this method should be possible by copying all the data to a new object, appendng the new content and writing back to S3 , but I don't think it's worth the effort in cases where the object data (file) is very large ( _I may be wrong_ ) . Please for your opinion.
> 
> [Reference: Append Data to S3 Object](https://stackoverflow.com/questions/41783903/append-data-to-an-s3-object)

Hello, 

Yes, I agree with you. Let's forget abt the append method. It is an optional method anyway.

Thank you so much.

Forchapeatl committed:
Merge branch 'Avaiga:develop' into Added-AWS-S3-DataNode

Forchapeatl committed:
configure_s3_data_node() method and  _STORAGE_TYPE_VALUE_S3 attribute

Forchapeatl committed:
Added test on S3ObjecctDataNode made from S3Oject DataNodeConfig

Forchapeatl commented:
Hello everyone, the requested changes have been made. This includes.

- [x]  Implement `configure_s3_object_data_node()` method to the `DataNodeConfig` class
- [x] `Add _STORAGE_TYPE_VALUE_S3` attribute and add that attribute to the `_ALL_STORAGE_TYPES` list.
- [x]  Add configuration tests for S3ObjectDataNode that created from an s3_object DataNodeConfig.
- [x] Update DataNodeConfig docstring.( _I don't really know what this means but i believe it's been covered_ ).

![image](https://github.com/Avaiga/taipy-core/assets/24577149/b7de8fd1-c1ea-49a5-8e44-1ddd7da7005a)

Please let me know if more changes are required.

trgiangdo commented on src/taipy/core/config/data_node_config.py:
```diff
@@ -1030,6 +1052,59 @@ def _configure_mongo_collection(
         )
 
+    @classmethod
+    def _configure_s3_object(
```
This method should also be injected to Config. You can add a line similar to other datanode types as:

https://github.com/Avaiga/taipy-core/blob/a01269fb268fe0001ca6ee27263b77271c53eb83/src/taipy/core/config/__init__.py#L39-L58

and


https://github.com/Avaiga/taipy-core/blob/626ef42c6bc584609134109331915e33c2f59f03/tests/conftest.py#L357-L375

trgiangdo commented on src/taipy/core/config/data_node_config.py:
```diff
@@ -1030,6 +1052,59 @@ def _configure_mongo_collection(
+        id: str,
+        aws_access_key: str,
+        aws_secret_acces_key: str,
+        aws_s3_bucket_name: str,
```
```suggestion
        aws_s3_bucket_name: str,
        aws_s3_object_key: str,
```

The `aws_s3_object_key` is missing right?

trgiangdo commented on src/taipy/core/config/data_node_config.py:
```diff
@@ -1030,6 +1052,59 @@ def _configure_mongo_collection(
+            id (str): The unique identifier of the new S3 Object data node configuration.
+            aws_access_key (str): Amazon Web Services ID for to identify account.
+            aws_secret_acces_key (str): Amazon Web Services access key to authenticate programmatic requests.
+            aws_s3_bucket_name (str): The bucket in S3 to read from and to write the data to.
```
Missing description of the `aws_s3_object_key`

Forchapeatl replied:
sorry about that

trgiangdo commented on tests/core/data/test_aws_s3_data_node.py:
```diff
@@ -0,0 +1,94 @@
+    @mock_s3
+    @pytest.mark.parametrize('data', [('Hello, other world!'),])
+    @pytest.mark.parametrize("properties", __properties)
+    def test_read(self, properties, data ):
```
This test is no different from the `test_write` method.

The test should do the following:
1. Connect to the bucket and write object to the bucket without using Taipy.
2. Create S3ObjectDataNode and try to read from the bucket.
3. Check if data read by S3ObjectDataNode is the correct data stored in the bucket.

If not, the test should just be renamed as `test_read_write`.

Forchapeatl replied:
@trgiangdo, this seems fair, the 3 step approach makes more sense, but the bucket will be declared as a Gobal variable , is this okay with you ?

trgiangdo replied:
why is it need to be a global variable?

Forchapeatl replied:
Your logic is right and straight forward. I don't know why I keep thinking of globals. Don't mid me :)

trgiangdo commented on tests/core/data/test_aws_s3_data_node.py:
```diff
@@ -0,0 +1,94 @@
+    @mock_s3
+    @pytest.mark.parametrize('data', [('Hello, world!'),])
+    @pytest.mark.parametrize("properties", __properties)
+    def test_write(self, properties, data ):
```
This test is no different from the `test_read` method.

The test should do the following:
1. Create S3ObjectDataNode and write data to the data node.
2. Connect to the bucket and read data using boto3.
3. Check if data stored in the bucket is the exact data that written to the S3ObjectDataNode.

If not, the test should just be renamed as `test_read_write`.

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,155 @@
+        edit_in_progress: bool = False,
+        editor_id: Optional[str] = None,
+        editor_expiration_date: Optional[datetime] = None,
+        properties: Dict = None,
```
```suggestion
        properties: Optional[Dict] = None,
```

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,155 @@
+        owner_id: Optional[str] = None,
+        parent_ids: Optional[Set[str]] = None,
+        last_edit_date: Optional[datetime] = None,
+        edits: List[Edit] = None,
```
```suggestion
        edits: Optional[List[Edit]] = None,
```

jrobinAV commented on src/taipy/core/data/aws_s3.py:
```diff
@@ -0,0 +1,155 @@
+                self.__AWS_SECRET_ACCESS_KEY,
+                self.__AWS_REGION,
+                self.__AWS_STORAGE_BUCKET_NAME,
+                self.__AWS_S3_OBJECT_PARAMETERS,
```
IS the `__AWS_S3_OBJECT_KEY` missing?

Forchapeatl replied:
sorry about that

Forchapeatl committed:
Merge branch 'Avaiga:develop' into Added-AWS-S3-DataNode

Forchapeatl committed:
Added configure_s3_object

Forchapeatl committed:
Added s3_object and config parameters

Forchapeatl committed:
Added s3_object key, s3_object to storagetype , switched to optional array

Forchapeatl committed:
added s3_object to string format

Forchapeatl committed:
changed storage type to s3_object

Forchapeatl committed:
Added configure_s3_object_data_node

Forchapeatl committed:
Added test_configure_s3_object_data_node

Forchapeatl committed:
Added test for s3_object_data_node configuration

Forchapeatl committed:
included missing s3_object_key

Forchapeatl committed:
test check for s3_object storage type

Forchapeatl committed:
Seperated test for s3_objects read write methods

 test_write method.

1. Connects to the bucket and write object to the bucket without using Taipy.
2. Creates S3ObjectDataNode and try to read from the bucket.
3. Checks if data read by S3ObjectDataNode is the correct data stored in the bucket.

test_read method

1. Creates S3ObjectDataNode and write data to the data node.
2. Connects to the bucket and read data using boto3.
3. Checks if data stored in the bucket is the exact data that written to the S3ObjectDataNode.

Forchapeatl committed:
Added S3ObjectDataNode

Forchapeatl commented:
![image](https://github.com/Avaiga/taipy-core/assets/24577149/e0cbe650-0f9c-4966-8c44-1172c4cde5b8)

![image](https://github.com/Avaiga/taipy-core/assets/24577149/85c3e273-92fb-48a6-af27-de2ddb3bd9c5)

```py
#tests3.py
import taipy as tp
import boto3
from taipy import Config, Core

def build_message(s3_object,name):
    return f"Hello {name}!"

s3_object_cfg = Config.configure_s3_object_data_node(
    id="test_s3_object",
    aws_access_key="ENTER YOUR AWS ACCESS_KEY",
    aws_secret_access_key="ENTER YOUR AWS ACCESS_SECRET_KEY",
    aws_s3_bucket_name="ENTER YOUR BUCKET NAME", #Must be Already existing bucket
    aws_s3_object_key="taipy_object")

input_name_data_node_cfg = Config.configure_data_node(id="input_name")
message_data_node_cfg = Config.configure_data_node(id="message")
build_msg_task_cfg = Config.configure_task("build_msg", build_message, input=[s3_object_cfg,input_name_data_node_cfg], output=[message_data_node_cfg])
scenario_cfg = Config.configure_scenario("scenario", task_configs=[build_msg_task_cfg])

if __name__ == "__main__":

    Core().run()
    hello_scenario = tp.create_scenario(scenario_cfg)
    myS3ObjectDataNode = hello_scenario.test_s3_object
    myS3ObjectDataNode.write("This is nice , Hello , HI , WHAT? ,WHEN! Who?")
    ans = myS3ObjectDataNode.read()
    print(ans)

    hello_scenario.input_name.write("Taipy")
    hello_scenario.submit()

```

Hello everyone, things seem to be woring fine now. The `S3ObjectDataNode` does get created. However it seems to be failing storage type test.

![image](https://github.com/Avaiga/taipy-core/assets/24577149/0edde470-90ae-4de4-9c81-9912c7333e57)

Is this a file or variable`caplog.text` ? Please for more information on this.
My tests failes on these lines:
https://github.com/Forchapeatl/taipy-core/blob/57b29061626186aa4839528f8eaf81baac15cf5e/tests/core/test_core.py#L35
https://github.com/Forchapeatl/taipy-core/blob/57b29061626186aa4839528f8eaf81baac15cf5e/tests/core/config/test_data_node_config.py#L132
https://github.com/Forchapeatl/taipy-core/blob/57b29061626186aa4839528f8eaf81baac15cf5e/tests/core/config/checkers/test_data_node_config_checker.py#L94

Please bear with me on the `tests3.py` code sample. The [docs ](https://docs.taipy.io/en/latest/manuals/core/entities/data-node-mgt/)seems advanced for me.

trgiangdo commented:
The failed tests are because you updated the error message with new storage_type, but the test `expected_error_message` on the tests are not updated.

A small update should fix the problem.

Forchapeatl commented:
> The failed tests are because you updated the error message with new storage_type, but the test `expected_error_message` on the tests are not updated.
> 
> A small update should fix the problem.

Wow, thankyou @trgiangdo ,  I managed to pass 2 more test,  any hints on [this file](https://github.com/Forchapeatl/taipy-core/blob/57b29061626186aa4839528f8eaf81baac15cf5e/tests/core/config/checkers/test_data_node_config_checker.py#L94) ?

trgiangdo commented:
I believe it has the same problem. A small update on the `expected_error_message` should fix it.

Forchapeatl committed:
Fixed assertion error on error_message for test_check_storage_type

Forchapeatl committed:
fixed error_message

Forchapeatl committed:
Added S3 object to error_message

Forchapeatl commented:
![image](https://github.com/Avaiga/taipy-core/assets/24577149/1562d8cb-a4d2-4478-8199-c062ff031b74)
Passed all 1108 tests. I belive s3ObjectDataNode has been succesfully added and tested.

trgiangdo commented:
The tests are failing:
- On [installation tests](https://github.com/Avaiga/taipy-core/actions/runs/7093338473/job/19309984255?pr=828):
```
File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/taipy/core/data/aws_s3.py", line 13, in <module>
    import boto3
ModuleNotFoundError: No module named 'boto3'
```
- On [backend tests](https://github.com/Avaiga/taipy-core/actions/runs/7093338538/job/19309986170?pr=828):
```
>       assert expected_error_message in caplog.text
E       assert '`storage_type` field of DataNodeConfig `data_nodes` must be either csv, sql_table, sql, mongo_collection, pickle, excel, generic, json, parquet, s3_object, or in_memory. Current value of property `storage_type` is "bar".' in 'ERROR    Taipy:config.py:237 `storage_type` field of DataNodeConfig `data_nodes` must be either csv, sql_table, sql, ..., pickle, excel, generic, json, parquet, in_memory, or s3_object. Current value of property `storage_type` is "bar".\n'
E        +  where 'ERROR    Taipy:config.py:237 `storage_type` field of DataNodeConfig `data_nodes` must be either csv, sql_table, sql, ..., pickle, excel, generic, json, parquet, in_memory, or s3_object. Current value of property `storage_type` is "bar".\n' = <_pytest.logging.LogCaptureFixture object at 0x118d6c970>.text
```
The `parquet, s3_object, or in_memory` is different from `parquet, in_memory, or s3_object`

Can you check again?

Forchapeatl committed:
fixed check_storage_type to assert error_message

Forchapeatl committed:
Added boto3 to requirements

jrobinAV commented on tests/conftest.py:
```diff
@@ -371,6 +371,8 @@ def init_config():
             ("configure_excel_data_node", DataNodeConfig._configure_excel),
             ("configure_generic_data_node", DataNodeConfig._configure_generic),
+            ("configure_s3_object_data_node", DataNodeConfig._configure_s3_object),
+
```
```suggestion
```

jrobinAV committed:
Update tests/conftest.py

remove blank line

trgiangdo commented:
The tests are up and running. That's great 😁

Can you take a look at the linter error in https://github.com/Avaiga/taipy-core/actions/runs/7102562575/job/19333215991?pr=828?
This is so that we have consistent code style in our project.

Forchapeatl committed:
Fixed linter error

Forchapeatl committed:
Fixed Linter error

Forchapeatl committed:
Fixed Linter errors

Forchapeatl committed:
Fixed Linter errors

Forchapeatl committed:
Fixed Linter errors

Forchapeatl committed:
Fixed linter errors

Forchapeatl committed:
removed trailing white spaces

Forchapeatl committed:
fixed linter errors

Forchapeatl committed:
Fixed syntax error

Forchapeatl committed:
Added new line

Forchapeatl commented:
>The tests are up and running. That's great 😁

Thank you. The code should be working fine now.

Forchapeatl commented:
Well :) this is unexpected , @trgiangdo , I did run the pylint test as you said . I am abit reluctant to fix [this error](https://github.com/Avaiga/taipy-core/actions/runs/7120008303/job/19405012106?pr=828) because , I am unable to reproduce the error on my environment , let me know what you think.
![image](https://github.com/Avaiga/taipy-core/assets/24577149/d30e4046-77bb-418c-9543-c062c1017052)

trgiangdo commented:
About the error
```
Running: pycodestyle --max-line-length=120 --ignore=E121,E123,E126,E226,E24,E704,W503,W504,E203 .
./tests/core/data/test_aws_s3_data_node.py:56:62: E231 missing whitespace after ','
./tests/core/data/test_aws_s3_data_node.py:76:61: E231 missing whitespace after ','
```
You just need to add a space after the comma in line 56 and line 76 and it should works

Forchapeatl committed:
Fixed pycodestyle error

trgiangdo commented:
Hello @Forchapeatl.

FYI, this taipy-core repository is merged into the taipy repository and is going to be archived soon.

So, instead of merging your PR, can you create a new PR in the main taipy repository, so you can be counted as an official Taipy contributor by Github?
The content of the PR should be very similar.

trgiangdo commented:
Migrated to https://github.com/Avaiga/taipy/pull/585
